File: frontend/src/lib/queues.ts
import { api, readErrorMessage } from "@/lib/api";

export interface SimpleQueue {
  id: string | null;
  name: string;
  target: string | null;
  max_limit: string | null;
  burst_limit: string | null;
  burst_threshold: string | null;
  burst_time: string | null;
  parent: string | null;
  priority: string | null;
  bytes_in: number | null;
  bytes_out: number | null;
  disabled: boolean;
  comment: string | null;
}

export interface SimpleQueueCreate {
  name: string;
  target?: string | null;
  max_limit?: string | null;
  burst_limit?: string | null;
  burst_threshold?: string | null;
  burst_time?: string | null;
  parent?: string | null;
  priority?: string | null;
  disabled?: boolean;
  comment?: string | null;
}

export async function listSimpleQueues(deviceId: string): Promise<SimpleQueue[]> {
  return api.get(`devices/${deviceId}/queues/simple`).json<SimpleQueue[]>();
}

export async function createSimpleQueue(
  deviceId: string,
  payload: SimpleQueueCreate,
): Promise<{ id: string }> {
  try {
    return await api
      .post(`devices/${deviceId}/queues/simple`, { json: payload })
      .json<{ id: string }>();
  } catch (e) {
    throw new Error(await readErrorMessage(e));
  }
}

export async function deleteSimpleQueue(deviceId: string, queueId: string): Promise<void> {
  try {
    await api.delete(`devices/${deviceId}/queues/simple/${encodeURIComponent(queueId)}`);
  } catch (e) {
    throw new Error(await readErrorMessage(e));
  }
}

export async function resetSimpleQueueCounters(
  deviceId: string,
  queueId: string,
): Promise<void> {
  try {
    await api.post(
      `devices/${deviceId}/queues/simple/${encodeURIComponent(queueId)}/reset-counters`,
    );
  } catch (e) {
    throw new Error(await readErrorMessage(e));
  }
}

File: backend/app/schemas/queues.py
from __future__ import annotations

from pydantic import BaseModel, Field


class SimpleQueuePublic(BaseModel):
    id: str | None
    name: str
    target: str | None
    max_limit: str | None
    burst_limit: str | None
    burst_threshold: str | None
    burst_time: str | None
    parent: str | None
    priority: str | None
    bytes_in: int | None
    bytes_out: int | None
    disabled: bool
    comment: str | None


class SimpleQueueCreate(BaseModel):
    name: str = Field(min_length=1, max_length=64)
    target: str | None = Field(default=None, max_length=255)
    max_limit: str | None = Field(
        default=None,
        max_length=64,
        description="upload/download in bps with suffixes, e.g. '10M/10M'",
    )
    burst_limit: str | None = Field(default=None, max_length=64)
    burst_threshold: str | None = Field(default=None, max_length=64)
    burst_time: str | None = Field(default=None, max_length=64)
    parent: str | None = Field(default=None, max_length=64)
    priority: str | None = Field(default=None, max_length=64)
    disabled: bool = False
    comment: str | None = Field(default=None, max_length=255)
